### PR TITLE
Implement customFunction to replace middlewareUtils

### PR DESCRIPTION
**Pull request by ianmacartney**

### CODE REVIEW

File: convex/lib/withSession.ts [hunk 1]
```diff
@@ -125,80 +38,48 @@ export const withOptionalSession = generateMiddleware<
...
-  MergeArgsForRegistered<Args, ObjectType<typeof sessionMiddlewareValidator>>,
-  Output
->;
-export function mutationWithSession(func: any): any {
-  return mutation(withSession(func));
-}
+export const mutationWithSession = customMutation(mutation, {
+  args: { sessionId: v.id("sessions") },
+  input: async (ctx, args) => {
+    const session = (await ctx.db.get(args.sessionId)) ?? null;
```
Comment (sshader) [inline convex/lib/withSession.ts:44]: Is it intentional that mutation requires a session but query does not?

File: convex/lib/withSession.ts [hunk 2]
```diff
@@ -125,80 +38,48 @@ export const withOptionalSession = generateMiddleware<
...
+    if (session === null) {
+      throw new Error(
+        "Session must be initialized first. " +
+          "Are you wrapping your code with <SessionProvider>? " +
+          "Are you requiring a session from a query that executes immediately?"
```
Comment (sshader) [inline convex/lib/withSession.ts:49]: This shouldn't say `query`, right? Maybe remove the line?

**Review by sshader — COMMENTED** [review]
I would be excited to see the READMEs / docs for how to use `customFunction` (could also potentially take a stab at this myself)

**Comment by ianmacartney** [comment]
> I would be excited to see the READMEs / docs for how to use customFunction (could also potentially take a stab at this myself)

README in progress!

**Pull request merged**